### ISSUE TITLE
another insidious bug with Game::remove_road_forward

### ISSUE THREAD

**Issue by tlongstretch** [freeserf/freeserf#521]
this is a likely cause for much agony. 

```
   if (map->has_serf(pos)) {
      Serf *serf = get_serf_at_pos(pos);
      if (!map->has_flag(pos)) {
        serf->set_lost_state();
```

When a path is deleted, any serf on the path becomes Lost.  This includes serfs that happen to be WORKING on the road, such as lumberjacks, foresters, farmers, stonecutters, etc.  This function should only apply to serfs that are walking-on or waiting-on the road, not to any serf found!   

The result is that the working/freewalking serf becomes Lost, but the building he is Holder to still thinks he is the Holder and so no replacement comes either, causing the building to cease operation.  This is itself a problem because in general, any time a serf becomes Lost, there should be some check that ensures any building he is Holder to has its Holder cleared and can request another serf.

I was able to reproduce this problem once I understood it by building a road through a forest, waiting until a lumberjack was chopping a tree while himself standing on a path, and I deleted the road.  The lumberjack became Lost, and his hut still showed him being Holder

**Comment by tlongstretch** [comment]
I am adding code to only make a serf on road lost if he is in one of the various walking/waiting/idle on road states, and also adding a function to clear the holder of a building when a serf is made lost

**Comment by tlongstretch** [comment]
to help debug these issues, I am adding a variable to Serf class object to store the index of any building the serf is Holder to, so it can be cross-referenced for these checks

**Comment by jarrodmoldrich** [comment]
Really enjoying the flurry of commits I've been seeing lately, especially this one.  Reminds me of a bug in the Amiga version, where if you placed a flag at the border with another kingdom that had straight road right at the edge, you could visit the flag page and join the other kingdom's roads to yours.

Once I discovered this I used it as a feature.  Whenever I had usurped the lands of a neighbouring kingdom to the point that they didn't have any castles anymore, I would make one of these bridges and the serfs would just head into my stores and castles and start doing random jobs.  I think I first noticed this when an AI did it to me, and I found a serfs from other kingdoms just pottering about in mine.  It was somewhat revealing about how the AI was written, how robust the algorithms were, and how permissive the data structures were.  Once a serf found a graph of roads with a flag that belonged to them, they lived on that graph of roads.  It was never anticipated that the graphs of two kingdoms would merge (or maybe the writers knew about the 'bug' and left it in there).

**Comment by tlongstretch** [comment]
the most shocking thing to me I have just recently discovered is that (I think) the original game had no pathfinding at all!  The novel road-building and flag system appears to be a way to avoid the CPU cycles required for actual pathfinding searches.  When debugging various FreeWalking serf issues and wondering why the FreeWalking logic was so random and inefficient I finally realized that the original game must have no pathfinder at all, and it was added in Freeserf only.

**Comment by jarrodmoldrich** [comment]
Yeah it was really frustrating to just watch the serfs not be able to find their way back home when they lost their path/building.  Or the strange ways they would approach another castle to battle.   OTOH, I sort of appreciate the ockham's razor approach to the AI, and it damn well scaled on those little 7/14mhz processors.

The road system is actually one of the things I appreciate about Settlers 1/2 over other incarnations and other 'god games' in general.  A good road system could make or break the functioning of a city.  Unfortunately the AI was never good at it.  Maybe if there is an API for freeserf, a RL AI could train on it and make a compelling AI opponent.  It seems like a difficult problem to solve.  It felt like making a functioning city took real care, planning, and a bit of luck.  It's very easy to become SNAFUed by running out of a crucial resource.

**Comment by tlongstretch** [comment]
I feel the same way, that anything after Settlers2 was junk because of the changes in resource & road management.

So, I spent the past 3-4 winters working on an improved AI,  I'm pretty satisfied with it now.  Most of the hard work was on efficient road creation.  I haven't been able to convince Freeserf maintainers to add it though.

**Comment by jarrodmoldrich** [comment]
Cool!  That's forkserf?  I'll have to give it a shot when I have a week to blow on a game of Settlers. A time which seems to be disappearing well into the horizon.

**Comment by tlongstretch** [comment]
@jarrodmoldrich see the original thread https://github.com/freeserf/freeserf/issues/431, and current state...https://github.com/forkserf/forkserf

**Comment by jarrodmoldrich** [comment]
I remember this seeing this thread a loooong time ago.  Good work!

In terms of upstreaming, maybe there's some merit in making a general AI API.  I don't know how much motivation there is for other people to make AIs, but it would be cool to have someone fork your AI, make changes, and compete with your AI in the same game.

But AFAICT the Settlers dev community isn't that large.  So I think there should be at least one generic AI that's part of Freeserf.  The length of the game doesn't really lend well to online play vs other people.

**Comment by tlongstretch** [comment]
I believe I could adapt my AI logic to work as a multplayer implementation where it just pays attention to the state of the game and takes actions via an API, yes.  This way it could play any of the current implementations (Freeserf, Freeserf.NET, Forkserf, Serflings)
One major hurdle is that nobody is developing in the same language.  Its C++, C#, and Java all competing.  But if they used the same API for multiplayer you could have an network AI opponent that is language agnostic and add them to your games

**Comment by jarrodmoldrich** [comment]
Wasn't aware of Seflings or Freeserf.Net.  Having an API might even cause some development efforts to converge on Freeserf.  But I understand this is a passion project, and I know for certain I don't have the bandwidth to take on a job like that, and wouldn't expect it of you – so this is all just fun conjecture :)  Thanks for your work with the AI.  I look forward to trying it out.

**Comment by tlongstretch** [comment]
I no longer see the "holder missing" issues after applying this fix.  I didn't put too much thought into the correct serf states to check for, but this is pretty good

```
    if (map->has_serf(pos)) {
      Serf *serf = get_serf_at_pos(pos);
      if (!map->has_flag(pos)) {
        Log::Debug["game"] << "considering calling set_lost_state on serf #" << serf->get_index() << " at pos " << pos << " case1, map says no flag here";
        // I think this is causing issues with free walking serfs that Hold a building, such as lumberjacks, rangers, farmers, stonecutters
        //  when they are on a road that gets removed, they become Lost and their building loses its Holder.  These serfs should NOT become
        //  Lost unless they are walking/waiting on the road.
        // here are the states I think are possible for "serf walking on a path to get somewhere"
        //  ALL OTHER STATES I assume are either serf is in a building or serf is free walking or otherwise doing something that isn't tied to a road
        if (serf->get_state() == Serf::StateWalking || serf->get_state() == Serf::StateTransporting || serf->get_state() == Serf::StateReadyToEnter
         || serf->get_state() == Serf::StateDelivering || serf->get_state() == Serf::StateIdleOnPath || serf->get_state() == Serf::StateWaitIdleOnPath 
         || serf->get_state() == Serf::StateWakeAtFlag || serf->get_state() == Serf::StateWakeOnPath ){
          Log::Debug["game"] << "serf #" << serf->get_index() << " at pos " << pos << " in a walking-on-a-path compatible state, calling set_lost_state on serf #" << serf->get_index() << " at pos " << pos << " case1, map says no flag here";
          serf->set_lost_state();
        }else{
          Log::Debug["game"] << "serf #" << serf->get_index() << " at pos " << pos << " is in state #" << serf->get_state() << " which is NOT requiring being on a road, NOT calling set_lost_state on serf #" << serf->get_index() << " at pos " << pos << " case1, map says no flag here";
        }

```
